feat(context): add release commit range

jdrst opened the pull request:
## Description

Adds the used `commit_range` for each release. 
`commit_range` includes the full length commit ids (`commit_range.from` and `commit_range.to`) and the short ids (`commit_range.from_short` and `commit_range.to_short`) 
Also added related docs (and documentation for the sort_commits default)

## Motivation and Context

See #1041

## How Has This Been Tested?

Via integration_test and three new fixtures.

## Types of Changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Documentation (no code change)
- [ ] Refactor (refactoring production code)
- [ ] Other

## Checklist:

- [ ] My code follows the code style of this project.
- [x] I have updated the documentation accordingly.
- [x] I have formatted the code with [rustfmt](https://github.com/rust-lang/rustfmt).
- [x] I checked the lints with [clippy](https://github.com/rust-lang/rust-clippy).
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.


/edit: added a third fixture. no more caveats + documentation and rustfmt/clippy

jdrst commented:
working on it :D

jdrst commented:
i am unsure how to fix that timeout. is this flaky?

orhun commented:
That links check should be fine... I'll take care of it later

orhun commented on website/docs/usage/adding-commits.md:
```diff
@@ -27,3 +27,5 @@ git commit -m "$commit_msg"
 
 The commit SHA will be empty as default when `--with-commit` is used. Specify the hash with a message separated by single whitespace for setting the commit SHA. e.g. `--with-commit "8f55e69eba6e6ce811ace32bd84cc82215673cb6 feat: add X"`
+
+Custom commits will have no effects on the commit range.
```
This is a bit unclear for the reader - maybe update it like:

```
:::info

 These added commits will have empty [`commit_range`](docs/etc/etc) variable etc.

::: 
```

Or something along those lines.

orhun commented on git-cliff-core/src/commit_range.rs:
I would love to rename this to `range` for brevity :)

orhun commented on git-cliff-core/src/commit_range.rs:
```diff
@@ -0,0 +1,27 @@
+
+use crate::commit::Commit;
+
+#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
+/// Commit range (from..to or from_short..to_short)
```
```suggestion
/// Commit range (from..to or from_short..to_short)
#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
```

orhun commented on git-cliff-core/src/commit_range.rs:
```diff
@@ -0,0 +1,27 @@
+#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
+/// Commit range (from..to or from_short..to_short)
+pub struct CommitRange {
+	from:       String,
+	to:         String,
+	from_short: String,
+	to_short:   String,
```
Can you add comments explaining what these fields are?

orhun commented on git-cliff-core/src/commit_range.rs:
```diff
@@ -0,0 +1,27 @@
+		Self {
+			from:       from.id.clone(),
+			to:         to.id.clone(),
+			from_short: from.short_id.clone(),
+			to_short:   to.short_id.clone(),
```
I'm thinking maybe we can remove the `short` variants... the templating can be used to trim them anyways.

e.g.

```
{{ commit_range.to | truncate(length=7, end="") }}
``` 

Also I have another idea which might change the whole implementation but... maybe adding this to the `commit` object would be cleaner?

e.g.

```
{{ commit.range.to | truncate(length=7, end="") }}
``` 

What do you think?

jdrst replied:
> I'm thinking maybe we can remove the short variants... the templating can be used to trim them anyways.

Short ids aren't necessarily unique when just truncated. I'd consider using the actual short id coming from git as more robust.

> Also I have another idea which might change the whole implementation but... maybe adding this to the commit object would be cleaner?

I'm not sure i can follow? The commit has no idea about a range, or does it? Or do you mean, instead of having the range on the release, have it on every commit in the release? I think i'd consider that counterintuitive.

orhun replied:
> Short ids aren't necessarily unique when just truncated. I'd consider using the actual short id coming from git as more robust.

I'd like to double down on this though... We don't have 2 variants (short, long) in the `commit` object itself and it might lead to inconsistencies (in very rare cases) if we go with short-long here. So I'd prefer we only go with the long variant here.

Thoughts?

jdrst replied:
I'd also like to double down :D (one last try). 
What kind of inconsistencies?
I'd rather change e.g. the commit in the context to be an object with long+short (that'd be breaking though).

orhun replied:
> What kind of inconsistencies?

You know, we'll be supporting short hashes in one variable (`commit_range`) but not for the other (`commit`) which is not great.

The reason why I don't like supporting short commit hashes is because 7 characters is not long enough and it's more probable that two entirely different commits might belong to the same short hash.

Even worse, you can override the first 7 digits of a commit hash with a tool like [`lucky-commit`](https://github.com/not-an-aardvark/lucky-commit):

```
$ git log
1f6383a Some commit

$ lucky_commit

$ git log
0000000 Some commit
```

So you can have a repo with all `000000` ids, see [this git history](https://github.com/not-an-aardvark/lucky-commit/commits/main/).

jdrst replied:
> You know, we'll be supporting short hashes in one variable (commit_range) but not for the other (commit) which is not great.

That i absolutely understand, hence my proposition to also add it to the commit itself. But one could also argue that `commit` and `commit_range` are different use cases with different needs. 

> The reason why I don't like supporting short commit hashes is because 7 characters is not long enough and it's more probable that two entirely different commits might belong to the same short hash.

That is exactly my point, just from a different perspective :D. This PR does make use of [git_object_short_id](https://libgit2.org/docs/reference/main/object/git_object_short_id.html) to circumvent that ([see here](https://github.com/jdrst/git-cliff/blob/1b8ecf8f3c527d5ce57f2538d36797231f21f1b6/git-cliff-core/src/commit.rs#L186))

Sorry for being so persistent. You can just put your foot down and say "Please remove the short id"!

Cyclonit replied:
I'll try playing tie-breaker ;)

I'd remove the short variants too. We are talking about configuration here. I don't see a point in ommitting a few bytes here. Putting in the long hash doesn't hurt and is more expressive.

Having only one set simplifies both the code and the documentation too.

orhun replied:
Hey again, needed to take some break due to a small injury but I'm hoping to get this merged before the release soon :)

So I think we should go without short hashes here and rebase on `main` to fix the conflicts. The rest should be ready.

@jdrst do you mind incorporating these changes? 😊

orhun reviewed (commented):
Thanks for the implementation!

I had some comments and ideas :)

orhun commented:
We need to rebase on `main` and fix conflicts 😊

jdrst commented:
Hey, no problem! I'm currently on vacation though. I could offer some time within the week after the next (so somewhere between 28.4 and 2.5). Does that still fit your timeframe? 
If not, i think you have editing rights, so if it's really pressing you could do it yourself. The rebase was uncomplicated, i just didn't push it yet :/


On April 20, 2025 9:48:24 PM GMT+02:00, "Orhun Parmaksız" ***@***.***> wrote:
>@orhun commented on this pull request.
>
>
>
>> +			from_short: from.short_id.clone(),
>+			to_short:   to.short_id.clone(),
>
>Hey again, needed to take some break due to a small injury but I'm hoping to get this merged before the release soon :)
>
>So I think we should go without short hashes here and rebase on `main` to fix the conflicts. The rest should be ready.
>
>@jdrst do you mind incorporating these changes? 😊 
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/orhun/git-cliff/pull/1123#discussion_r2051794578
>You are receiving this because you were mentioned.
>
>Message ID: ***@***.***>

orhun commented:
Hey that works, I also tried to do changes here but I can't push to your branch :/

```
 ! [remote rejected]   jdrst/main -> jdrst/main (permission denied)
error: failed to push some refs to 'https://github.com/jdrst/git-cliff.git'
```

orhun commented on git-cliff-core/src/range.rs:
We can move the entirety to the `commit` module I think.

orhun commented on git-cliff/src/lib.rs:
```diff
@@ -397,6 +398,22 @@ fn process_repository<'a>(
 		}
 	}
 
+	// Set the commit ranges for all releases
+	for release in &mut releases {
```
Is it possible that we use an already existing loop for this instead of looping through releases again?

e.g. we can set this in

https://github.com/orhun/git-cliff/blob/aec41befc701e786c01ec4ba5399c67df1981779/git-cliff/src/lib.rs#L302

Maybe

jdrst replied:
Will have a look. From what i remember it would've needed a lot of branching/complicated logic because we don't know all of the commits (or the release) we're in until everything is added (and then we can only add it before we switch to the previous release).

I think i opted for this because it is O(n) with n being the number of releases). That should still be fine even with lots of releases.

lehmanju replied:
Have a look at `process_submodules`. It does the same thing because it also needs to have the first and last commit of a release. Maybe you can merge the two loops?

jdrst commented:
> Hey that works, I also tried to do changes here but I can't push to your branch :/
> 
> ```
>  ! [remote rejected]   jdrst/main -> jdrst/main (permission denied)
> error: failed to push some refs to 'https://github.com/jdrst/git-cliff.git'
> ```

That is weird..
![image](https://github.com/user-attachments/assets/9a9b328c-bf84-461d-894a-7be1abe78e15)

I just turned it off and on again. Maybe it broke because i accidentally force pushed no changes in between and this PR got closed automatically?

orhun commented:
No I didn't work :/ Maybe you have branch protection on on `main` or something, I don't really know...

jdrst commented:
> No it didn't work :/ Maybe you have branch protection on on `main` or something, I don't really know...

no idea.. i just forked. no branch protection active. the hint even states 

> If checked, users with write access to orhun/git-cliff can add new commits to your main branch.
> You can always change this setting later.

i just added you as contributor..

orhun commented:
Ok I just pushed a commit to there... it doesn't show up in this PR though. What is going on...

orhun commented:
Ok I messed up something

orhun commented:
Can we maybe create a new branch and open a new PR from there? Creating a PR from the `main` branch makes things super confusing...

jdrst commented:
> Can we maybe create a new branch and open a new PR from there? Creating a PR from the `main` branch makes things super confusing...

aight, will do next week. you could upload a patch here or smth.

orhun commented:
The changes are in the `main` branch here: https://github.com/jdrst/git-cliff/tree/main

jdrst commented:
> The changes are in the `main` branch here: https://github.com/jdrst/git-cliff/tree/main

currently rebasing, cleaning up and incorporating requested changes, will pr soon.
i squashed your commit into mine and added you as co-author. i hope that's okay?

orhun commented:
yup! thanks
